### DISCUSSION TITLE
urql usage with suspense and `requestPolicyExchange`

### DISCUSSION THREAD

**Discussion by xPapla** [urql-graphql/urql#2695] [Q&A]
I'm trying to use urql in react with client-side suspense, but I'm struggling to make it work with auto `request-policy` upgrade.

Here I created a [sandbox](https://codesandbox.io/s/urql-suspense-request-policy-exchange-p6587m?file=/src/App.js) where you can check the code that I am running. I expect that whenever I render `Page1` again, the `requestPolicyExchange` should upgrade the request policy to `cache-and-network` to get fresh data, but when re-render happens I am still getting the old results from cache and there is no network request being made.

https://user-images.githubusercontent.com/6313377/191745152-90622cdd-6811-4c7b-bb13-8bf8562da4f0.mp4

This behavior changes to what I expected at the beginning, when I drop the `suspense: true` as you can see on the video below:

https://user-images.githubusercontent.com/6313377/191746231-56f1b8d0-de81-423a-bfca-36e1dd1ecf4a.mp4

Should I do something else in my setup to add support for suspense and `requestPolicyExchange`?

**Comment by JoviDeCroock** [comment]
Hmm that makes sense, it's probably hitting the required append-only cache for react suspense rather than re-issuing the operation to the urql client. This happens [here](https://github.com/FormidableLabs/urql/blob/main/packages/react-urql/src/hooks/useQuery.ts#L106) this cache is basically required to make these suspense/concurrent paradigms functional 😅

**Reply by xPapla** [reply]
Thanks for your answer. I went through the code that people posted on react related issues and came up with this as a workaround. I changed the default policy to be `cache-and-network` and went with downgrading the requests rather than upgrading them. I thought it would be great if `requestPolicyExchange` had some kind of possibility to manipulate with the cache that you linked to. Possibly TTL could be implemented there?

```tsx
import { Exchange, makeOperation, Operation, OperationResult } from "urql";
import { map, pipe, tap } from "wonka";

interface Options {
  ttlMs: number;
}

export const requestPolicyExchange =
  ({ ttlMs }: Options): Exchange =>
  ({ forward }) => {
    const operations = new Map<number, number>();

    const processIncomingOperation = (operation: Operation): Operation => {
      if (operation.kind !== "query" || operation.context.requestPolicy !== "cache-and-network") {
        return operation;
      }

      const currentTime = new Date().getTime();
      const lastOccurrence = operations.get(operation.key) ?? 0;
      const shouldDowngrade = currentTime - lastOccurrence <= ttlMs;

      if (shouldDowngrade) {
        return makeOperation(operation.kind, operation, {
          ...operation.context,
          requestPolicy: "cache-first",
        });
      }

      return operation;
    };

    const processIncomingResults = (result: OperationResult): void => {
      const meta = result.operation.context.meta;
      const isMiss = !meta || meta.cacheOutcome === "miss";

      if (isMiss) {
        operations.set(result.operation.key, new Date().getTime());
      }
    };

    return (ops$) =>
      pipe(forward(pipe(ops$, map(processIncomingOperation))), tap(processIncomingResults));
  };
```

**Reply by JoviDeCroock** [reply]
Yeah, the cache is currently based on query lifecycles. For example if your `Todos` component mounts a `useQuery` hook as soon as `Todos` is unmounted or changes variables we will dispose of that cache-key so I reckon the scenario we run into here is because you have two simultaneous identical queries mounted leading to the `cache-and-network` not applying. The thing is if we check `cache-and-network` in that piece of code then we would do too many requests when you unsuspend 😅
